File: Java/src/trees_graphs/InOrderTraversal.java
package trees_graphs;

import java.util.ArrayList;
import java.util.Stack;

public class InOrderTraversal {
	public static ArrayList<Integer> traverseWithRecursion(TNode<Integer> root){
		if(root == null) return new ArrayList<Integer>();
		ArrayList<Integer> ret = new ArrayList<Integer>();
		ret.addAll(traverseWithRecursion(root.left));
		ret.add(root.data);
		ret.addAll(traverseWithRecursion(root.right));
		return ret;
	}
	public static ArrayList<Integer> traverseWithoutRecursion(TNode<Integer> root){
		Stack<TNode<Integer>> st = new Stack<TNode<Integer>>();
		ArrayList<Integer> ret = new ArrayList<Integer>();
		TNode<Integer> cur = root;
		while(cur.left != null){
			st.push(cur);
			cur= cur.left;
		}
		while(true){
			ret.add(cur.data);
			if(cur.right != null) {
				cur= cur.right;
				while(cur.left != null){
					st.push(cur);
					cur= cur.left;
				}
				ret.add(cur.data);
			}
			if(st.isEmpty()) break;
			else cur = st.pop();
		}
		return ret;
		
	}
}

File: CSharp/HelloUniverse/InOrderTraversal.cs
using System;
using System.Collections.Generic;

namespace HelloUniverse
{
    public static class InOrderTraversal
    {
        public class TreeNode
        {
            public int val;
            public TreeNode left;
            public TreeNode right;
            public TreeNode(int x) 
            { 
                val = x; 
            }
        }

        public static List<int> InorderTraversalIterative(TreeNode root)
        {
            var result = new List<int>();
            var st = new Stack<TreeNode>();
            var cur = root;
            while(st.Count != 0 || cur != null)
            {
                while(cur != null)
                {
                    st.Push(cur);
                    cur = cur.left;
                }

                cur = st.Pop();
                result.Add(cur.val);
                cur = cur.right;
            }

            return result;
        }

        public static List<int> InorderTraversalRecursive(TreeNode root)
        {
            if(root == null)
            {
                return new List<int>();
            }

            var left = InorderTraversalRecursive(root.left);
            var right = InorderTraversalRecursive(root.right);

            left.Add(root.val);
            left.AddRange(right);

            return left;
        }
    }
}
